keps/sig-apps/3998-job-success-completion-policy/README.md

```diff
@@ -0,0 +1,910 @@
+	// represented as "1,3-5,7".
+	// When this field is null, this field doesn't default to any value
+	// and is never evaluated at any time.
+	// 
+	// +optional
+	SucceededIndexes *string
+
+	// SucceededCount specifies the minimal required size of the actual set of the succeeded indexes
+	// for the job. When SucceededCount is used along with SucceededIndexes, the check is 
+	// constrained only to the set of indexes specified by SucceededIndexes.
+	// When this field is null, this field doesn't default to any value
+	// and is never evaluated at any time.
+	// When specified it needs to be a positive integer.
+	//
+	// +optional
+	SucceededCount *int
+}
+...
+
+// These are valid conditions of a job.
+const (
+	// JobSuccessCriteriaMet means the job has been succeeded.
+	JobSucceessCriteriaMet JobConditionType = "SuccessCriteriaMet"
+	...
+)
+```
+
+Moreover, we validate the following constraints for the `criteria`:
+- whether the specified indexes in the `succeededIndexes` and
+  the number of indexes in the `succeededCount` don't exceed the value of `completions`.
+- whether `Indexed` is specified in the `completionMode` field.
+- whether the size of `succeededIndexes` is under 64Ki.
+- whether the `succeededIndexes` field has a valid format.
+- whether the `succeededCount` field has an absolute number.
+- whether the criteria haven't changed.
+- whether the successPolicies meet the `succeededCount <= |succeededIndexes|`, 
+where `|succeededIndexes|` means the number of indexes in the `succeededIndexes`. 
+
+In addition, when the `successPolicy` isn't nil, we validate the `criteria` isn't nil.
+
+### Evaluation
+
+Every time the pod condition are updated, the job-controller evaluates the successPolicies following the rules in order:
+
+- `succeededIndexes`: the job-controller evaluates `.status.completedIndexes` to see if a set of indexes is there.
+- `succeededCount`: the job-controller evaluates `.status.succeeded` to see if the value is `succeededCount` or more.
+
+After that, the job-controller adds a `SuccessCriteriaMet` condition instead of a `Failed` condition to `.status.conditions`
+and the job-controller terminates the lingering pods. At that time, `JobSuccessPolicy` is set to the `status.reason` field.
+
+Note that when the job meets one of successPolicies, other successPolicies are ignored.
+
+Finally, once all pods have terminated, the job-controller adds a `Complete` condition to `.status.conditions`.
+
+If any successPolicy isn't set, the job-controller adds both a `SuccessCriteriaMet` and a `Complete` conditions to the Job
+after the job finished.
+
+### Transition of "status.conditions"
+
+When the job with successPolicies is submitted, the job `status.conditions` transits in the following:
+Note that the Job doesn't have an actual `Running` condition in the `status.conditions`.
+
+```mermaid
+stateDiagram-v2
+    [*] --> Running
+    Running --> Failed: Exceeded backoffLimit\n or Matched FailurePolicy
+    Failed --> [*]
```

Comment:
What about FailureTarget? Today, when we get to FailureTarget, we wouldn't transition to Complete.

is that correct  @mimowo?

Reply:
That's right. IIUC, when the Job meets the FailurePoliy with FailJob action, the Job will transition to Failed. In others, the Job wouldn't transition to Failed.

Let me rephrase here.

Reply:
Done.

Reply:
It doesn't transition to Failed directly. It first transitions to FailureTarget.
Only once all the other pods terminate, it transitions to Failed.

Reply:
That makes sense. Thak you for the clarifications!

Reply:
@alculquicondor @mimowo I updated this diagram. Please let me know if this is not correct.

Reply:
Thanks for the diagram update, it mostly look good. 

I'm only surprised by:
`FailureTarget --> SuccessCriteriaMet: Matched SuccessPolicy`. 

I would expect that once `FailureTarget` is added, then `Failed` is the only possible fate, and `SuccessCriteriaMet` is not added, we just skip evaluating the success policy a this point. Similarly, when `SuccessCriteriaMet` is added, then the only possible fate is "Complete". So, the two conditions `FailureTarget` and `SuccessCriteriaMet` are "symmetric".

Reply:
> I would expect that once FailureTarget is added, then Failed is the only possible fate, and SuccessCriteriaMet is not added, we just skip evaluating the success policy a this point. Similarly, when SuccessCriteriaMet is added, then the only possible fate is "Complete". So, the two conditions FailureTarget and SuccessCriteriaMet are "symmetric".

@mimowo IIUC, the `FailureTarget` condition is added to the Job even though the matched failurePolicy has `action=Ignore`,  `action=FailIndex`, or `action=Count`, right? So, I think that when the job meets failurePolicy with those actions, that means that the job failed.

Reply:
`FailureTarget --> Failed: Exceeded backoffLimit\nor action=FailJob`

So, when the job meets the `FailureTarget` with `action=FailJob`, successPolicy isn't evaluated.

Reply:
> @mimowo IIUC, the FailureTarget condition is added to the Job even though the matched failurePolicy has action=Ignore, action=FailIndex, or action=Count, right? 

`FailureTarget` is added when FailJob action is matched. It is used to await for all pods to terminate before setting `Failed: true`.

I'm sure how to read this transition: "FailureTarget --> SuccessCriteriaMet" in the diagram. 

Does it mean that we can have `FailureTarget: true`, then  `SuccessCriteriaMet: true`? I think this transition should be prohibited. Once FailureTarget: true we skip evaluation of success policy, don't we?

Reply:
> FailureTarget is added when FailJob action is matched. It is used to await for all pods to terminate before setting Failed: true.

@mimowo Oh, I see. It seems that my understanding is not correct. If so, we need to remove this transition, `FailureTarget --> SuccessCriteriaMet` as you said.

Reply:
@mimowo I updated the diagram here: https://github.com/kubernetes/enhancements/pull/4062/commits/5511faca093867e97cf629239b3a99a925268e7d

Thank you for pointing this out.

Reply:
> > FailureTarget is added when FailJob action is matched. It is used to await for all pods to terminate before setting Failed: true.
>
> @mimowo Oh, I see. It seems that my understanding is not correct. 

Here is the code reference where we generate the condition when FailJob action is matched: https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/job/job_controller.go#L814-L817

Reply:
@mimowo One more quetion. Is this transition possible, `FailureTarget --> Failed`?
I'm wondering if this transition never happenes. It means that I'm wondering if the transition of `FailureTarget --> [*]` is correct.

Reply:
> @mimowo I updated the diagram here: https://github.com/kubernetes/enhancements/commit/5511faca093867e97cf629239b3a99a925268e7d

Thanks! lgtm

Reply:
> > > FailureTarget is added when FailJob action is matched. It is used to await for all pods to terminate before setting Failed: true.
> > 
> > 
> > @mimowo Oh, I see. It seems that my understanding is not correct.
> 
> Here is the code reference where we generate the condition when FailJob action is matched: https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/job/job_controller.go#L814-L817

Thank you for the code! It would be very helpful!

Reply:
@mimowo Aldo, could you check this comment?

> @mimowo One more quetion. Is this transition possible, FailureTarget --> Failed?
> I'm wondering if this transition never happenes. It means that I'm wondering if the transition of FailureTarget --> [*] is correct.

https://github.com/kubernetes/enhancements/pull/4062#discussion_r1481288697

Reply:
I think `FailureTarget --> [*] ` isn't correct, we always transition `FailureTarget -> Failed`. Failed is the terminal state.

Reply:
> I think `FailureTarget --> [*] ` isn't correct, we always transition `FailureTarget -> Failed`. Failed is the terminal state.

It makes sense. The job-controller has only 2 terminal states, "Complete" and "Failed".
Thank you for letting me know :)

Reply:
The diagram lgtm currently